### PR TITLE
Create MySQL cluster with primary-replica setting

### PR DESCRIPTION
**Pull request by kfyharukz**
This PR implements the feature to bootstrap MySQL cluster with primary-replica setting. Moco configures `mysql-0` as primary and the rests as replicas, and set the index of current primary `MySQLCluster.status.currentPrimaryIndex`. 

The design document is [here](https://neco-wiki.netlify.app/research_mysql_detail_design/).

### CODE REVIEW

**Comment by ysksuzuki** [comment]
@ymmt2005 Please review the following.

- [mysql_clustering.go](https://github.com/cybozu-go/moco/pull/50/files#diff-c48da1ae4a7966b2ee06382c8e42bac5)
  - Reconcile MySQL Cluster with decideNextOps functions referring to the pattern CKE applies
- [mysql_accessor.go](https://github.com/cybozu-go/moco/pull/50/files#diff-c4dbf55b361ed6aff97cdd11a85821ee)
  - Handling database connections using sqlx library

File: constants.go [hunk 1]
```diff
@@ -96,3 +112,24 @@ const (
...
 	MiscPasswordKey = "MISC_PASSWORD"
 )
+
+const (
+	// InitializedClusterIndexField is an index name for Initialized MySQL Clusters
+	InitializedClusterIndexField = ".status.conditions.type.initialized"
+)
+
+const (
+	OperatorUpdatePrimary        = "update-primary"
```
Comment (ysksuzuki) [inline constants.go:122]: Comments are missing

File: controllers/mysql_clustering.go [hunk 2]
```diff
@@ -0,0 +1,664 @@
...
+}
+
+// TODO: Implementation for failover
+func selectPrimary(status *MySQLClusterStatus, cluster *mocov1alpha1.MySQLCluster) int {
+	return 0
+}
+
+func updatePrimary(cluster *mocov1alpha1.MySQLCluster, newPrimaryIndex int) []Operator {
+	currentPrimaryIndex := cluster.Status.CurrentPrimaryIndex
+	if currentPrimaryIndex != nil && *currentPrimaryIndex == newPrimaryIndex {
```
Comment (ysksuzuki) [inline controllers/mysql_clustering.go:475]: updatePrimary doesn't change `rpl_semi_sync_master_wait_for_slave_count` when the replicas field is changed from 3 to 5.

File: controllers/mysql_clustering.go [hunk 3]
```diff
@@ -0,0 +1,664 @@
...
+	secret := &corev1.Secret{}
+	myNS, mySecretName := getSecretNameForController(cluster)
+	err := i.Get(ctx, client.ObjectKey{Namespace: myNS, Name: mySecretName}, secret)
+	if err != nil {
+		return "", err
+	}
+	return string(secret.Data[passwordKey]), nil
+}
+
+type updatePrimaryOp struct {
```
Comment (kfyharukz) [inline controllers/mysql_clustering.go:521]: Split file which only includes Ops.

File: docs/crd_mysql_cluster.md [hunk 4]
```diff
@@ -51,13 +51,22 @@ MySQLClusterStatus
...
+| `reason`             | string | No       | One-word CamelCase reason for the condition's last transition.             |
+| `message`            | string | No       | Human-readable message indicating details about last transition.           |
+| `lastTransitionTime` | [Time] | Yes      | The last time the condition transits from one status to another.           |
+
+MySQLClusterConditionType
+------------------------
+
+| Value       | Description                                                                   |
+| ----------- | ----------------------------------------------------------------------------- |
+| Initialized | All the objects needed to make up the cluster have been created successfully. |
```
Comment (ymmt2005) [inline docs/crd_mysql_cluster.md:67]: Didn't you add more values?

File: controllers/mysql_clustering_test.go [hunk 5]
```diff
@@ -0,0 +1,521 @@
...
+)
+
+const (
+	CLUSTER   = "test-cluster"
+	NAMESPACE = "test-namespace"
+	UID       = "test-uid"
+	REPLICAS  = 3
+)
+
+func Test_decideNextOperation(t *testing.T) {
```
Comment (kfyharukz) [inline controllers/mysql_clustering_test.go:26]: Rename

**Review by ymmt2005 — CHANGES_REQUESTED** [review]

File: runners/cluster_watcher.go [hunk 6]
```diff
@@ -0,0 +1,69 @@
+package runners
```
Comment (kfyharukz) [inline runners/cluster_watcher.go:1]: Should we write a test?

File: controllers/mysql_accessor.go [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+	err := db.Ping()
+	if err != nil {
+		delete(acc.dbs, uri)
+		return nil, err
+	}
+	return db, nil
+}
+
+func (acc *MySQLAccessor) getURI(host, user, password string) string {
+	return fmt.Sprintf("%s:%s@tcp(%s:%d)/?timeout=%.0fs&readTimeout=%.0fs&interpolateParams=true", user, password, host,
```
[the remark this answers is not in the document]
Comment (zoetrope) [reply controllers/mysql_accessor.go:64]: https://pkg.go.dev/github.com/go-sql-driver/mysql?tab=doc#Config

File: constants.go (hunk 1 above)
Comment (kfyharukz) [reply constants.go:122]: Move these consts to a new file of Ops.

File: controllers/mysql_clustering.go (hunk 2 above)
Comment (kfyharukz) [reply controllers/mysql_clustering.go:475]: We will do it in the EPIC Enrich functions.

File: controllers/mysql_clustering.go [hunk 8]
```diff
@@ -0,0 +1,664 @@
...
+	ops = acceptWriteRequest(status, cluster)
+	if len(ops) != 0 {
+		return &Operation{
+			Conditions:     availableCondition(outOfSyncInts),
+			Operators:      ops,
+			SyncedReplicas: &syncedReplicas,
+		}, nil
+	}
+
+	return &Operation{
```
[the remark this answers is not in the document]
Comment (kfyharukz) [reply controllers/mysql_clustering.go:166]: Update API call with no changes makes no actual call.  So, it isn't needed.

**Review by zoetrope — DISMISSED** [review]
LGTM

File: Makefile [hunk 9]
```diff
@@ -20,15 +20,16 @@ GOBIN=$(shell go env GOBIN)
...
+MYSQL_VERSION := 8.0.21
 
 all: build/moco-controller
 
 # Run tests
 test:
 	cd /tmp; GO111MODULE=on GOFLAGS= go install github.com/cybozu/neco-containers/golang/analyzer/cmd/custom-checker
 	test -z "$$(gofmt -s -l . | grep -v '^vendor' | tee /dev/stderr)"
-	test -z "$$(golint $$(go list -mod=vendor ./... | grep -v /vendor/) | tee /dev/stderr)"
+	test -z "$$(staticcheck ./... | tee /dev/stderr)"
```
Comment (ymmt2005) [inline Makefile:32]:
```suggestion
	staticcheck ./...
```

**Review by ymmt2005 — APPROVED** [review]
Looks great!

**Review by zoetrope — APPROVED** [review]

**Pull request merged**